### ISSUE TITLE
Does the `--fla` supports backward pass of state gradients?  

### ISSUE THREAD

**Issue by shouldsee** [JL-er/RWKV-PEFT#11]
Thanks for sharing this great project. As per README, does it imply that the --fla flag could correctly differentiate through the states without bptt truncation?

> infctx train
"--train_type infctx --chunk_ctx 512" "chunk_ctx" represents the chunk length, while "ctx_len" stands for the total length of the data. Due to the lack of gradients in the wkv6state operator, I now recommend using fla instead.

---

The --fla flag seems to be 4x slower than CUDA module

**Comment by shouldsee** [comment]
BTW I am a bit confused with the --train_type infctx behaviour, especially unable to capture the expected state tuning effect with `--chunk_ctx 4096 --ctx_len 4096`

**Comment by JL-er** [comment]
> Thanks for sharing this great project. As per README, does it imply that the --fla flag could correctly differentiate through the states without bptt truncation?
> 
> > infctx train
> > "--train_type infctx --chunk_ctx 512" "chunk_ctx" represents the chunk length, while "ctx_len" stands for the total length of the data. Due to the lack of gradients in the wkv6state operator, I now recommend using fla instead.
> 
> The --fla flag seems to be 4x slower than CUDA module

Yes, FLA can be used independently, and CUDA operators will be faster when bsz is greater than 8.FLA is more suitable for state tuning, although there are still some accuracy issues being fixed (but they do not affect normal use).

**Comment by JL-er** [comment]
> BTW I am a bit confused with the --train_type infctx behaviour, especially unable to capture the expected state tuning effect with `--train_type infctx --chunk_ctx 4096 --ctx_len 4096`
> 
> Shouldnt the infctx state tuning similar to `--train_type state --ctx_len 4096`?

State tuning is not yet supported under infctx.If you want to use state tuning for training, you don't need chunk_ctx and truncation.

**Comment by JL-er** [comment]
> > BTW I am a bit confused with the --train_type infctx behaviour, especially unable to capture the expected state tuning effect with `--train_type infctx --chunk_ctx 4096 --ctx_len 4096`
> > Shouldnt the infctx state tuning similar to `--train_type state --ctx_len 4096`?
> 
> State tuning is not yet supported under infctx.If you want to use state tuning for training, you don't need chunk_ctx and truncation.

However, state tuning will be enabled with infctx in the future (infctx is just a method to save memory, trading time for space).

**Comment by shouldsee** [comment]
> > > BTW I am a bit confused with the --train_type infctx behaviour, especially unable to capture the expected state tuning effect with `--train_type infctx --chunk_ctx 4096 --ctx_len 4096`
> > > Shouldnt the infctx state tuning similar to `--train_type state --ctx_len 4096`?
> > 
> > 
> > State tuning is not yet supported under infctx.If you want to use state tuning for training, you don't need chunk_ctx and truncation.
> 
> However, state tuning will be enabled with infctx in the future (infctx is just a method to save memory, trading time for space).

Thanks. BTW has the infctx method been tested to be same as fixed-length method? My attempt seems to indicate some discrepancy between infctx and fixed length

**Comment by shouldsee** [comment]
After further testing, the --infctx mode with --fla is working fine. The discrepancy probably came from fitting both shift_state and wkv_state, instead of wkv_state only.

Fitting wkv_state only leads to faster convergence, as indicated by the lower line

![image](https://github.com/user-attachments/assets/a651a3f3-1cca-4f33-b518-4e4a1cac051f)

**Comment by JL-er** [comment]
> After further testing, the discrepancy might have come from difference between CUDA and fla
> 
> Using CUDA instead of --fla leads to faster convergence, as indicated by the lower line. But this discrepancy seems also affected by other hyperparameters...
> 
> ![image](https://private-user-images.githubusercontent.com/8034156/354632639-a651a3f3-1cca-4f33-b518-4e4a1cac051f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjI4NDc0NzQsIm5iZiI6MTcyMjg0NzE3NCwicGF0aCI6Ii84MDM0MTU2LzM1NDYzMjYzOS1hNjUxYTNmMy0xY2NhLTRmMzMtYjUxOC00ZTRhMWNhYzA1MWYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDgwNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA4MDVUMDgzOTM0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NTg4ZDhjOGJmNzBkYzBlYzQyNzg4YzVlZDNmODBjZWFhYmQ0NmQ4NDI0ODU1ZmE1MTY5ZjY3NDBhMWNjODBiYiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.HvMSR0q8qbiH7i1yVAXrMkbMkQZ9WtWTNdatq2qvf_Q)

Yes, the state tuning error of CUDA operators is larger than that of FLA, and it needs to be fixed.

**Comment by JL-er** [comment]
The latest fla has fixed accuracy issues。
